src/system/SystemObject.h

```diff
@@ -204,9 +197,78 @@ class ObjectPool
 public:
     void Reset();
 
-    T * TryCreate();
     void GetStatistics(chip::System::Stats::count_t & aNumInUse, chip::System::Stats::count_t & aHighWatermark);
 
+    template <typename... Args>
+    T * TryCreate(Args &&... args)
+    {
+        T * lReturn = nullptr;
+        static_assert(std::is_base_of<Object, T>::value, "Try to create an object which is not derived from class Object");
+
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+        T * newNode = new T(std::forward<Args>(args)...);
+
+        if (newNode->TryCreate())
+        {
+            std::lock_guard<std::mutex> lock(mMutex);
+            Object * p = &mDummyHead;
+            if (p->mNext)
+            {
+                p->mNext->mPrev = newNode;
+            }
+            newNode->mNext     = p->mNext;
+            p->mNext           = newNode;
+            newNode->mPrev     = p;
+            newNode->mMutexRef = &mMutex;
+            lReturn            = newNode;
+        }
+        else
+        {
+            delete newNode;
+        }
+#else // CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+        unsigned int lIndex = 0;
+
+        for (lIndex = 0; lIndex < N; ++lIndex)
```

Comment:
Similarly here, it seems like we should see if we can out-of-line and de-templatize as much as we can here (e.g. have a non-templated non-inline method to find and TryCreate the slot and return a pointer to put the new object into, etc).